lib/middleware.js

```diff
@@ -31,3 +35,16 @@ module.exports.applyMiddleware = function (argv, middlewares) {
       }
     }, argv)
 }
+
+module.exports.applyPreCheckMiddlware = function (argv, middlewares, context, aliases) {
+  for (let i = 0; i < middlewares.length; i++) {
+    let preChecksContext = {
+      'commands': JSON.parse(JSON.stringify(context.commands)),
+      'availableOptions': JSON.parse(JSON.stringify(aliases))
```

Comment:
do we apply aliases after _middleware_ is run? I can understand why this makes writing the middleware annoying for your use-case @trevorlinton.

What if the second argument to middleware was `yargs`, and we expose a `getAliases` method on `yargs`. _then_, if we can figure out an API for passing `applyBeforeValidation` to command specific middleware, I think we'd have addressed your use-cases without the need for this `context` object.

CC: @aorinevo,  @Khaledgarbaya